src/main/java/org/generationcp/middleware/service/impl/user/UserServiceImpl.java

```diff
@@ -201,26 +217,58 @@ public Integer updateUser(final UserDto userDto) {
 			user.setCloseDate(currentDate);
 			user.setStatus(userDto.getStatus());
 
-			// update user roles to the particular user
-			final UserRole role = user.getRoles().get(0);
-			if (!role.getRole().equals(userDto.getRole())) {
-				role.setRole(userDto.getRole());
+			final List<UserRole> userRoles = new ArrayList<>();
+			if (userDto.getUserRoles() != null) {
+				for (final UserRoleDto userRoleDto : userDto.getUserRoles()) {
+					boolean found = false;
+					for (final UserRole userRole : user.getRoles()) {
+						if (userRole.getRole().getId().equals(userRoleDto.getRole().getId()) &&
+							(userRole.getCropType() == null && userRoleDto.getCrop() == null || userRole.getCropType().getCropName()
+								.equals(userRoleDto.getCrop().getCropName())) &&
+							(userRole.getWorkbenchProject() == null && userRoleDto.getProgram() == null || userRole.getWorkbenchProject()
+								.getUniqueID().equals(userRoleDto.getProgram().getUuid()))) {
+							userRoles.add(userRole);
+							found = true;
+							break;
+						}
+					}
+					if (!found) {
+						final UserRole userRole = new UserRole();
```

Comment:
The clause inside `!found `in `updateUser `is duplicated in `createUser`. We should consolidate to one method.